### PR TITLE
go/mysql: performance optimizations in protocol encoding

### PR DESCRIPTION
**Pull request by mattrobenolt**
This employs a couple tricks that combined seemed fruitful:

* Swapping to binary.LittleEndian.Put* on the basic calls gets us a free
  boost while removing code. The main win from this swap is the slice
  boundary check, resulting in a massive boost. I kept it inlined, but
  added my own boundary checking in `writeLenEncInt` since swapping it
  out here resulted in a very minor performance regression from the
  current results. I assume from the extra coersion needed to the uint*
  type, and another reslice.
* Reslicing the byte slice early so all future operations work on
  0-index rather than pos+ indexing. This seemed to be a pretty sizeable
  win without needing to do more addition on every operation later to
  determine the index, they get swapped out for constants.
* Read path employs the same early reslicing, but already has explicit
  bounds checks.
* Rewrite `writeZeroes` to utilize the Go memclr optimization.

```
$ benchstat {old,new}.txt
goos: darwin
goarch: arm64
pkg: vitess.io/vitess/go/mysql
                                 │    old.txt     │               new.txt                │
                                 │     sec/op     │    sec/op     vs base                │
EncWriteInt/16-bit-10               0.4685n ±  0%   0.3516n ± 0%  -24.94% (p=0.000 n=10)
EncWriteInt/16-bit-lenencoded-10     2.049n ±  0%    2.049n ± 0%        ~ (p=0.972 n=10)
EncWriteInt/24-bit-lenencoded-10     1.987n ±  0%    2.056n ± 0%   +3.45% (p=0.000 n=10)
EncWriteInt/32-bit-10               0.7819n ±  0%   0.3906n ± 0%  -50.05% (p=0.000 n=10)
EncWriteInt/64-bit-10               1.4080n ±  0%   0.4684n ± 0%  -66.73% (p=0.000 n=10)
EncWriteInt/64-bit-lenencoded-10     3.126n ±  0%    2.051n ± 0%  -34.40% (p=0.000 n=10)
EncWriteZeroes/4-bytes-10           2.5030n ±  0%   0.3123n ± 0%  -87.52% (p=0.000 n=10)
EncWriteZeroes/10-bytes-10          4.3815n ±  0%   0.3120n ± 0%  -92.88% (p=0.000 n=10)
EncWriteZeroes/23-bytes-10          8.4575n ±  0%   0.3124n ± 0%  -96.31% (p=0.000 n=10)
EncWriteZeroes/55-bytes-10         20.8750n ± 10%   0.6245n ± 0%  -97.01%
EncReadInt/16-bit-10                 2.050n ±  0%    2.068n ± 1%   +0.90% (p=0.001 n=10)
EncReadInt/24-bit-10                 2.034n ±  0%    2.050n ± 0%   +0.76% (p=0.000 n=10)
EncReadInt/64-bit-10                 2.819n ±  1%    2.187n ± 0%  -22.41% (p=0.000 n=10)
geomean                              2.500n         0.8363n       -66.55%
```

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/encoding.go [hunk 1]
```diff
@@ -137,10 +132,111 @@ func writeLenEncString(data []byte, pos int, value string) int {
 }
 
 func writeZeroes(data []byte, pos int, len int) int {
```
[the remark this answers is not in the document]
Comment (dbussink) [reply go/mysql/encoding.go:134]: Let's remove it then.

**Comment by dbussink** [comment]
You'll need to update the commits and force push to ensure the DCO sign off.

**Comment by dbussink** [comment]
@mattrobenolt Is clearing this way faster than using the following?

```go
func writeZeroes(data []byte, pos int, len int) int {
	data = data[pos : pos+len]
	for i := range data {
		data[i] = 0
	}
	return pos + len
}
```

The Go compiler recognizes that pattern and optimizes it into a `CALL runtime.memclrNoHeapPointers(SB)` essentially. Is that just as fast or even faster? See also https://github.com/golang/go/issues/5373 for that pattern.

**Comment by mattrobenolt** [comment]
@dbussink lol so

```go
package main

import "testing"

func writeZeroesVitessMain(data []byte, pos, len int) int {
	for i := 0; i < len; i++ {
		data[pos+i] = 0
	}
	return pos + len
}

func writeZeroesSpecialized23(data []byte, pos int) int {
	data = data[pos:]

	_ = data[22]
	data[0] = 0
	data[1] = 0
	data[2] = 0
	data[3] = 0
	data[4] = 0
	data[5] = 0
	data[6] = 0
	data[7] = 0
	data[8] = 0
	data[9] = 0
	data[10] = 0
	data[11] = 0
	data[12] = 0
	data[13] = 0
	data[14] = 0
	data[15] = 0
	data[16] = 0
	data[17] = 0
	data[18] = 0
	data[19] = 0
	data[20] = 0
	data[21] = 0
	data[22] = 0

	return pos + 23
}

func writeZeroesSpecialized10(data []byte, pos int) int {
	data = data[pos:]

	_ = data[9]
	data[0] = 0
	data[1] = 0
	data[2] = 0
	data[3] = 0
	data[4] = 0
	data[5] = 0
	data[6] = 0
	data[7] = 0
	data[8] = 0
	data[9] = 0

	return pos + 10
}

func writeZeroesMemclr(data []byte, pos, len int) int {
	end := pos + len
	data = data[pos:end]

	for i := range data {
		data[i] = 0
	}

	return end
}

func BenchmarkZeroes(b *testing.B) {
	buf := make([]byte, 128)

	b.Run("vitess-main/23-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesVitessMain(buf, 16, 23)
		}
	})

	b.Run("vitess-main/10-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesVitessMain(buf, 16, 10)
		}
	})

	b.Run("specialized/23-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesSpecialized23(buf, 16)
		}
	})

	b.Run("specialized/10-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesSpecialized10(buf, 16)
		}
	})

	b.Run("memclr/23-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesMemclr(buf, 16, 23)
		}
	})

	b.Run("memclr/10-byte", func(b *testing.B) {
		for range b.N {
			_ = writeZeroesMemclr(buf, 16, 10)
		}
	})
}
```

I put them all side by side and this is wild.

```
$ go test -v . -bench=.
goos: darwin
goarch: arm64
pkg: x
BenchmarkZeroes
BenchmarkZeroes/vitess-main/23-byte
BenchmarkZeroes/vitess-main/23-byte-10          138833698                8.634 ns/op
BenchmarkZeroes/vitess-main/10-byte
BenchmarkZeroes/vitess-main/10-byte-10          268413115                4.471 ns/op
BenchmarkZeroes/specialized/23-byte
BenchmarkZeroes/specialized/23-byte-10          570586390                2.125 ns/op
BenchmarkZeroes/specialized/10-byte
BenchmarkZeroes/specialized/10-byte-10          1000000000               0.6531 ns/op
BenchmarkZeroes/memclr/23-byte
BenchmarkZeroes/memclr/23-byte-10               1000000000               0.3263 ns/op
BenchmarkZeroes/memclr/10-byte
BenchmarkZeroes/memclr/10-byte-10               1000000000               0.3286 ns/op
PASS
ok      x       7.020s
```

That memclr optimization is really good.

Going to swap that out, which will let me get rid of the specialized versions. I didn't like that anyways.

File: go/mysql/encoding.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattrobenolt) [reply go/mysql/encoding.go:134]: Irrelevant now that I swapped to the non-specialized-but-faster memclr variant.

**Comment by mattrobenolt** [comment]
@dbussink updated to the memclr optimization as well as benchmarks in the PR description.

**Review by dbussink — APPROVED** [review]

**Review by arthurschreiber — APPROVED** [review]

**Review by systay — APPROVED** [review]

**Comment by deepthi** [comment]
Nice work!

**Pull request merged**